src/include/processor/operator/persistent/index_builder_error_handler.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include "common/exception/copy.h"
+#include "common/types/types.h"
+#include "processor/execution_context.h"
+
+namespace kuzu {
+namespace storage {
+class NodeTable;
+}
+
+namespace processor {
+template<typename T>
+struct IndexBuilderError {
+    std::string message;
+    T key;
+    common::nodeID_t nodeID;
+};
+
+class IndexBuilderErrorHandler {
+public:
+    IndexBuilderErrorHandler(ExecutionContext* context, common::LogicalTypeID pkType,
+        storage::NodeTable* nodeTable, uint64_t queryID,
+        std::atomic<common::row_idx_t>* sharedDeletedRowCounter);
+
+    template<typename T>
+    void handleOrStoreError(IndexBuilderError<T> error) {
+        if (!ignoreErrors) {
+            throw common::CopyException(error.message);
+        }
+
+        if (getNumErrors() >= warningLimit) {
+            flushStoredErrors();
+        }
+
+        addNewVectors();
+        keyVector.back()->setValue<T>(0, error.key);
+        offsetVector.back()->setValue(0, error.nodeID);
+        errorMessages.emplace_back(std::move(error.message));
+    }
+
+    void flushStoredErrors();
+
+private:
+    static constexpr common::idx_t DELETE_VECTOR_SIZE = 1;
+    static constexpr uint64_t LOCAL_WARNING_LIMIT = 1024;
+
+    bool ignoreErrors;
+    uint64_t warningLimit;
+    ExecutionContext* context;
+    common::LogicalTypeID pkType;
+    storage::NodeTable* nodeTable;
+    uint64_t queryID;
+    std::atomic<common::row_idx_t>* sharedDeletedRowCounter;
```

Comment:
I don't enjoy passing atomic pointer around like this. Can u make pass the shared_ptr? Also I don't think we need `atomic` here, grabbing the lock if you can seems better to me.

Reply:
updated